runelite-client/src/main/java/net/runelite/client/plugins/runenergy/RunEnergyOverlay.java

```diff
@@ -0,0 +1,226 @@
+				"Run Time Remaining: " + getEstimatedRunTimeRemaining() + "</br>" +
+				"Recover Time Remaining: " + getEstimatedRecoverTimeRemaining();
+
+			tooltipManager.add(new Tooltip(tooltip));
+		}
+
+		return new Dimension((int) bounds.getWidth(), (int) bounds.getHeight());
+	}
+
+	private boolean isStaminaEffectActive()
+	{
+		final Widget[] runOrbChildren = client.getWidget(WidgetInfo.MINIMAP_RUN_ORB).getStaticChildren();
+
+		if (runOrbChildren == null)
+		{
+			return false;
+		}
+
+		// Check for the stamina effect sprite
+		for (final Widget child : runOrbChildren)
+		{
+			if (child.getSpriteId() == SpriteID.MINIMAP_ORB_RUN_ICON_SLOWED_DEPLETION)
+			{
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private String getEstimatedRunTimeRemaining()
+	{
+		// Calculate the amount of energy lost every 2 ticks (0.6 seconds).
+		// Negative weight has the same depletion effect as 0 kg.
+		final int effectiveWeight = Math.max(client.getWeight(), 0);
+		double lossRate = (Math.min(effectiveWeight, 64) / 100.0) + 0.64;
+
+		if (isStaminaEffectActive())
+		{
+			// Stamina effect reduces energy depletion to 30%
+			lossRate *= 0.3;
+		}
+
+		// Calculate the number of seconds left
+		final double secondsLeft = (client.getEnergy() * 0.6) / lossRate;
+		final int minutes = (int) Math.floor(secondsLeft / 60.0);
+		final int seconds = (int) Math.floor(secondsLeft - (minutes * 60.0));
+
+		// Return the text
+		return Integer.toString(minutes) + ":" + StringUtils.leftPad(Integer.toString(seconds), 2, "0");
+	}
+
+	private boolean isWearingFullGraceful()
+	{
+		final ItemContainer equipment = client.getItemContainer(InventoryID.EQUIPMENT);
+
+		if (equipment == null)
+		{
+			return false;
+		}
+
+		final Item[] items = equipment.getItems();
+
+		// Check that the local player is wearing enough items to be using full Graceful
+		// (the Graceful boots will have the highest slot index in the worn set).
+		if (items == null || items.length <= EquipmentInventorySlot.BOOTS.getSlotIdx())
+		{
+			return false;
+		}
+
+		return (ALL_GRACEFUL_HOODS.contains(items[EquipmentInventorySlot.HEAD.getSlotIdx()].getId()) &&
+			ALL_GRACEFUL_TOPS.contains(items[EquipmentInventorySlot.BODY.getSlotIdx()].getId()) &&
+			ALL_GRACEFUL_LEGS.contains(items[EquipmentInventorySlot.LEGS.getSlotIdx()].getId()) &&
+			ALL_GRACEFUL_GLOVES.contains(items[EquipmentInventorySlot.GLOVES.getSlotIdx()].getId()) &&
+			ALL_GRACEFUL_BOOTS.contains(items[EquipmentInventorySlot.BOOTS.getSlotIdx()].getId()) &&
+			ALL_GRACEFUL_CAPES.contains(items[EquipmentInventorySlot.CAPE.getSlotIdx()].getId()));
+	}
+
+	private String getEstimatedRecoverTimeRemaining()
+	{
+		// Calculate the amount of energy recovered every second
+		double recoverRate = (48 + client.getBoostedSkillLevel(Skill.AGILITY)) / 360.0;
+
+		if (isWearingFullGraceful())
+		{
+			// 30% recover rate increase from Graceful set effect
+			recoverRate *= 1.3;
```

Comment:
Graceful doesn't work like this. It only applies the effect if you are standing still

Reply:
Cheers. I'll have a look in a sec. :)

Reply:
If I understand correctly, run energy only ever regenerates when the player's not running anyway (and not busy doing something such as crossing agility obstacles).

In this case, maybe it would be useful to hide the recovery time while running?